python/cudf/cudf/core/_internals/where.py

```diff
@@ -0,0 +1,395 @@
+        ):
+            input_col = frame._data[frame.name]
+            return _check_and_cast_columns_with_scalar(
+                source_col=frame._data[frame.name],
+                other_scalar=other,
+                inplace=inplace,
+            )
+
+        elif isinstance(frame, DataFrame):
+            if cudf.utils.dtypes.is_scalar(other):
+                other = [other for i in range(len(frame._data.names))]
+
+            source_df = frame.copy(deep=False)
+            others = []
+            for col_name, other_sclr in zip(frame._data.names, other):
+
+                (
+                    source_col,
+                    other_scalar,
+                ) = _check_and_cast_columns_with_scalar(
+                    source_col=source_df._data[col_name],
+                    other_scalar=other_sclr,
+                    inplace=inplace,
+                )
+                source_df._data[col_name] = source_col
+                others.append(other_scalar)
+            return source_df, others
+        else:
+            raise ValueError(
+                f"Inappropriate input {type(frame)} "
+                f"and other {type(other)} combination"
+            )
+
+
+def where(
+    frame: Union[Series, Index, DataFrame],
+    cond: Any,
+    other: Any = None,
+    inplace: bool = False,
+) -> Optional[Union[Frame]]:
+    """
+    Replace values where the condition is False.
+
+    Parameters
+    ----------
+    cond : bool Series/DataFrame, array-like
+        Where cond is True, keep the original value.
+        Where False, replace with corresponding value from other.
+        Callables are not supported.
+    other: scalar, list of scalars, Series/DataFrame
+        Entries where cond is False are replaced with
+        corresponding value from other. Callables are not
+        supported. Default is None.
+
+        DataFrame expects only Scalar or array like with scalars or
+        dataframe with same dimension as frame.
+
+        Series expects only scalar or series like with same length
+    inplace : bool, default False
+        Whether to perform the operation in place on the data.
+
+    Returns
+    -------
+    Same type as caller
+
+    Examples
+    --------
+    >>> import cudf
+    >>> df = DataFrame({"A":[1, 4, 5], "B":[3, 5, 8]})
+    >>> df.where(df % 2 == 0, [-1, -1])
+        A  B
+    0 -1 -1
+    1  4 -1
+    2 -1  8
+
+    >>> ser = Series([4, 3, 2, 1, 0])
+    >>> ser.where(ser > 2, 10)
+    0     4
+    1     3
+    2    10
+    3    10
+    4    10
+    dtype: int64
+    >>> ser.where(ser > 2)
+    0       4
+    1       3
+    2    <NA>
+    3    <NA>
+    4    <NA>
+    dtype: int64
+    """
+
+    if isinstance(frame, DataFrame):
+        if hasattr(cond, "__cuda_array_interface__"):
+            cond = DataFrame(
+                cond, columns=frame._data.names, index=frame.index
+            )
+        elif not isinstance(cond, DataFrame):
+            cond = frame.from_pandas(pd.DataFrame(cond))
```

Comment:
What if `cond` is a `MultiIndex`?

Reply:
Added a shape check that will handle this case.